x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -212,6 +220,24 @@ public ApiKeyService(Settings settings, Clock clock, Client client, XPackLicense
             this.apiKeyAuthCache = CacheBuilder.<String, ListenableFuture<CachedApiKeyHashResult>>builder()
                 .setExpireAfterAccess(ttl)
                 .setMaximumWeight(maximumWeight)
+                .removalListener(notification -> {
+                    if (RemovalReason.EVICTED == notification.getRemovalReason() && getApiKeyAuthCache().count() >= maximumWeight) {
+                        evictionCounter.increment();
+                        logger.trace("API key with ID [{}] was evicted from the authentication cache, " +
+                                "possibly due to cache size limit", notification.getKey());
+                        final long previousCheck = lastEvictionCheckedAt.get();
+                        final long now = clock.instant().toEpochMilli();
```

Comment:
Should this use `System.nanoTime()` instead of epoch time?

Reply:
I thought `clock.instant().toEpochMilli()` would be faster than `System.nanoTime` because I thought the former is only millisecond accuracy thus not requiring extra read from hardware. But it turned out that `clock.instant()` already does the read for nanos. So it should have no performance advantage. I replaced it with `System.nanoTime`.